### DISCUSSION TITLE
How to use immich API for dummies?

### DISCUSSION THREAD

**Discussion by aisbergde** [immich-app/immich#6228] [Q&A]
### The feature

immich administrators can automate external libraries to be rescanned in the GUI. This is great, but it is limited.
On the other side, the API has needed functionality:
https://immich.app/docs/api/scan-library/
contains some code, how to refreshAllFiles
https://immich.app/docs/api/remove-offline-files

I want to call this and automate. But how?
There are cURL examples, but I have no idea, what to do and how to use them. I even don't know what cURL is.

Could this be explained in the docs for dummies like me? Some first steps, how to start
Or maybe some links to existing docs. Just to get a point, where to start. I have no idea, how to use the API at all. https://immich.app/docs/api/introduction is empty, and the next page starts with code examples. 

And I would like to know if these library commands can be used also with the default library, especially to remove offline files.
How to mark assets as offline, I found a manual way from the maintenance page to export JSON over a script into the database. I can't automate, but at least it is a way.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by aviv926** [comment]
I can try to explain about the use of Curl here, but I think it is less relevant because as you mentioned it is something that should be in the documentation.
In fact I would be happy to document everything the thing is I still don't understand everything 100% so it requires help from the development team, I guess in Discord or after there is a PR (even a draft) it will be easier to start making things happen.

**Reply by aisbergde** [reply]
At least a starting point would be fine. I don't know where to start. It looks like, there are so many possibilities. I am fine to start with databases, because they are my daily business. But I am a typical Windows guy, using GUI and so on :-)

**Reply by jrasm91** [reply]
We have an http API, so using it implies some level of understanding around http requests and responses. If you aren't a programmer or have no experience with sending and receiving http requests, educating you on those topics is a bit out of scope for Immich and our documentation pages. Something we _could_ do, is publish python and typescript modules/packages that would make it easier to use. Something like "install python" and then "copy paste this code". But, again, writing code still implies some level of programming or technical background.

**Reply by aisbergde** [reply]
I have some experience. I used the Service Now API to access it, using Phyton, to get data and to import into SQL Server. But the Service Now API is well documented and I found Python scripts, I could use and adapt. There was a starting point and I could start and solved all requirements.

Here I have no idea, where even to start. And I think, it should be even easier to use the API, if I don't want to import data, but if I only want to trigger some actions.

The main issue, normally I work on Windows. But my Windows knowledge is not helpful here, because immich is running on docker on my Synology. Synology is running an Unix / Linux

So, I am a developer for decades, but I am a BI developer, and I am restricted to Windows, because I use Microsoft software, which requires Windows: Visual Studio, SQL Server Management Studio. Even, if I wanted to switch from Windows, I couldn't.

I think, there should be some cron jobs, and these cron jobs need to call the API from time to to time.

**Reply by aviv926** [reply]
You can view example of python code by using example section:
![image](https://github.com/immich-app/immich/assets/51673860/0940163d-34a3-4453-a18f-2ed97370796e)

**Reply by MysterSoldier** [reply]
We definitely need a step by step for beginners. I meen like break it down Barney level. 1st day of 6th grade type step by step

**Reply by Dearone13** [reply]
Where I can find this section with examples?

**Comment by aisbergde** [comment]
My desired API usage job is currently limited to the question:
How to setup a cron job inside the immich installation to rescan libraries using the API, if the GUI accessible cron job doesn't work or doesn't do enough?

**Reply by martabal** [reply]
FYI, automatic scanning does not work but should be back in the next release

**Comment by aisbergde** [comment]
I changed it from feature request into Q&A. To avoid it to be closed.
Still hope to learn, how to use the API.

**Reply by aviv926** [reply]
You can use your ip address in order to use the API via Curl you just need to be connected to Immich to be able to use the API.

for example:
http://192.168.1.230:2283/api/asset

**Reply by aisbergde** [reply]
OK, this works. I can use it in the browser to get a result, for example
http://192.168.178.170:2283/api/user
http://192.168.178.170:2283/api/asset
http://192.168.178.170:2283/api/person

But now how to use curl?

reading the https://curl.se/docs/tutorial.html, in all examples the address is included. And I think, I need to provide the Access Token for authoizaion. I got the access token, but how to use it?

I try something in Powershell
This kind of command returns something
```
curl http://192.168.178.170:2283
``` 
But trying this, I get an error, somehow I need to provide the token
```
PS C:\Users\goert> curl http://192.168.178.170:2283/api/user
curl : {"message":"Authentication required","error":"Unauthorized","statusCode":401}
```
or
```
PS C:\Users\goert> curl http://192.168.178.170:2283/api/user -H 'Authorization: Bearer xxxxxmytoken'
Invoke-WebRequest : Der Parameter "Headers" kann nicht gebunden werden. Der Wert "Authorization: Bearer
xxxxxmytoken" vom Typ "System.String" kann nicht in den Typ
"System.Collections.IDictionary" konvertiert werden.
```

I also tried to connect via SSH to my Synology and to run curl there. And of course, I need authorization somehow.

Finally, I want to call something like
```
curl -L -X POST '/api/library/whateverid_i_should_findout/removeOffline'
```

And best would be to do this once a day, setting somehow a cron job inside the immich docker environment

**Reply by Zelnes** [reply]
I've been trying to figure it out, and finally found it.

You need to use the header `x-api-key`, like this (in `bash`):
```bash
api_key=xxxYYYzzz
curl 'https://immich.home/api/user?isAll=true' -H 'Accept: application/json' -H "x-api-key: $api_key"
```

I haven't seen it before, but there's actually a button to demonstrate how it works on the API documentation...
![immich-api-key](https://github.com/immich-app/immich/assets/19836933/9955fe6b-8f41-4241-813e-0dee43c8bc9d)

**Reply by melhakim** [reply]
This needs to go into the documentation. I could not find a pointer to how to set the api key except here.

**Reply by TechSupport26** [reply]
Based on what has been shown here I realised that I had to change the url to include the name of the API key.

http://192.168.1.10:2283/api_key_name/

Some context. I was setting up Heimdall and wanted to get the API for Immich to work.
Thanks.

**Reply by alextran1502** [reply]
@TechSupport26 if you are talking about connecting on your mobile app it is literal `http://localip:2283/api` not api key

**Reply by TechSupport26** [reply]
Perhaps my comment is not in the right place? No I am not talking about my mobile app.

I was just referring to the url including the name of the API that was made. Perhaps saying api_key_name is not accurate?
Perhaps I am wrong as Heimdall is not showing any additional information with the API in place.

> not api key

I mean the name of the API key not the key itself.

Maybe I am still missing something.

**Reply by bo0tzz** [reply]
It should just literally be the text `api`

**Reply by TechSupport26** [reply]
I see.

When I use the name of the key, I just get html
What I needed was http://localip:2283/api/api-keys or something like that.
Heimdall still does not work but that is probably not for here.

Thanks for repeating yourselves.

**Comment by njobrien1006** [comment]
I wasn't able to find documentation either (how to authorize with an API key).

Appears you need another header item `x-api-key` with a user's API key.

Example Windows Curl:
```
curl -L "192.168.1.245:30041/api/assets/bulk-upload-check" -H "Content-Type: application/json" -H "Accept: application/json" -H "x-api-key: YoUrReAlYLonGUserAPIKeYFroMWeBui" -d "{\"assets\": [{\"checksum\": \"string\", \"id\": \"string\"}]}"
```

Edit..........the documentation does have the info. However, is not displayed by default and you must hit `Authorize` at [API Docs](https://immich.app/docs/api/upload-asset/) for it to appear.

**Comment by seajean** [comment]
Has the same issue, I wanna call the api with the api key to get the album and image data, but don't want to dive info the immich code, I can not find a way to do this

**Reply by jrasm91** [reply]
You can add the query param `?apiKey=<key>` or the header `x-api-key: <key>`

**Reply by Zelnes** [reply]
Have a look at [this answer](https://github.com/immich-app/immich/discussions/6228#discussioncomment-8254671)

**Comment by profucius** [comment]
For anyone struggling to get a API call working to trigger the library scan, here is what I did:

First you need to create an API key by clicking your user icon at the top-right. Then click API Keys, and create a new one. Name it whatever, I named mine Automations. Copy the key to a Notepad.

In that same Notepad, a few lines down, paste this, then move the API key to where it says "APIKEYHERE"

`curl -L -X GET "192.168.1.x:PORT/api/libraries" -H "Content-Type: application/json" -H "Accept: application/json" -H "x-api-key: APIKEYHERE"`

Change the x in the IP address to your Immich IP address. Change the PORT to whatever the port number is.

Then open a CMD window on your Windows computer (if using CURL). Copy and paste the line from Notepad and hit enter to run it.

You'll get a response with a bunch of code looking stuff. The very first one looks like "id":"LONGIDCODE". You want to copy the LONGIDCODE from here, between the quote marks. 

If you only created one library in your Immich, you should see only one ID. If you created multiple libraries, then you'll get multiple IDs. When you trigger a scan via the API, you have to do this for each library/ID individually. So the below steps, you may have to do it multiple times if you have multiple libraries.

Copy the ID(s) from above into your Notepad.

`curl -L -X POST "192.168.1.x:PORT/api/libraries/IDCODE/scan" -H "Content-Type: application/json" -H "Accept: application/json" -H "x-api-key: APIKEYHERE"`

Change the x and PORT and APIKEYHERE again, like you did in the first command. This time you also need to change the IDCODE to the new ID you just pasted into Notepad.

Now copy that whole line and paste into the CMD window and hit enter. It should call Immich and trigger the library scan. If it worked, your command result wont show any responses. Now go into Immich>Jobs and refresh the page, you should see it scanning Active.

If this worked, you're good to go! You can create a .bat file with the contents of your last command and save it to the desktop or somewhere else. Now you can just double click this file, and it will trigger Immich to update immediately. If you had multiple libraries, just put each of these commands on a new line and change the IDs in each one (leave all the API keys etc the same). Running the file will trigger all the libraries at once.

This is a crude guide, there are many ways you can clean this up and automate it further. For example I am using an automation script to dynamically get the IDs of each library and run the commands for them, within a deeper photo syncing script that I tie into Syncthing. The sky's the limit.

**Comment by cristihcd** [comment]
Here is an example for calling the api from powershell:

- step 1 (once)
save the api key in an enrypted file

```powershell
$secureToken = Read-Host "Please enter your api key" -AsSecureString
$secureToken | Export-CliXml -Path "c:\some_path\api.xml"
```

Invoke-ApiScan.ps1
```powershell
# --- Configuration ---
$tokenFilePath = "c:\some_path\api.xml"
$libraryId = "your_library_id"
$apiUrl = "https://your_ip/api/libraries/$libraryId/scan"

try {
    # 1. Import the encrypted token from the file.
    $secureToken = Import-CliXml -Path $tokenFilePath

    # 2. Convert the SecureString back to plain text for the API call.
    $ptr = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($secureToken)
    $plainTextToken = [System.Runtime.InteropServices.Marshal]::PtrToStringBSTR($ptr)
    [System.Runtime.InteropServices.Marshal]::ZeroFreeBSTR($ptr)

    # 3. Prepare the headers using the correct x-api-key format.
    $headers = @{
        "Accept" = "application/json"
        "x-api-key" = $plainTextToken
    }

    Write-Host "Triggering library scan for ID: $libraryId..."
    
    # Added -SkipCertificateCheck for local HTTPS with a self-signed certificate.
    Invoke-RestMethod -Uri $apiUrl -Method Post -Headers $headers -SkipCertificateCheck

    Write-Host "API call successful!"

} catch {
    Write-Error "An error occurred: $_"
}

```

**Comment by shuhaowu** [comment]
I have this script to stack pixel phone raw files:

https://gist.github.com/shuhaowu/e315d74bdd444824338764d281e2af78

**Comment by timonrieger** [comment]
for anyone still looking for help, I maintain [immichpy](https://immichpy.timonrieger.de/), which provides a complete CLI for the Immich API, making it more convenient to interact with it than curl :)
